libs/scalalib/src/mill/scalalib/JavaModule.scala

```diff
@@ -647,7 +653,18 @@ trait JavaModule
     // (it's respectively provided, runtime, import). The configuration is compile for
     // standard mvnDeps / moduleDeps.
     //
-    JavaModule.InternalRepo(transitiveCoursierProjects().distinctBy(_.module.name.value))
+    val project = coursierProject()
+    // Mark optional direct dependencies as non-optional, so that these are included in the
+    // class paths of this module
+    val project0 = project.withDependencies0(
+      project.dependencies0.map {
+        case (conf, dep) if dep.optional =>
+          (conf, dep.withOptional(false))
+        case other =>
+          other
+      }
+    )
+    JavaModule.InternalRepo(Seq(project0) ++ transitiveCoursierProjects())
```

Comment:
I'm a bit surprised, we need to manually manage the handling of direct optional dependencies. Isn't this something that also must work in `sbt` and `coursier` cli tool?